@here/harp-mapview/lib/Utils.ts

```diff
@@ -242,53 +243,190 @@ export namespace MapViewUtils {
+        // read the comment above `maxTilt` below to understand why this is just an approximation.
+        let angleBetweenNormals = 0;
+        if (projection === sphereProjection) {
+            const rotationTargetNormal = projection.surfaceNormal(
+                rotationTargetWorld,
+                cache.vector3[0]
+            );
+            const mapTargetNormal = projection.surfaceNormal(mapTargetWorld, cache.vector3[1]);
+            angleBetweenNormals = rotationTargetNormal.angleTo(mapTargetNormal);
+        }
+
+        const ninetyRad = THREE.MathUtils.degToRad(90);
+        // This calculates the angle between the direction vector to the rotation center from the
+        // camera and the direction vector to the map view center (offsetX and offsetY == 0). This
+        // value stays constant when tilting but using the same X & Y offsets. We use the triangle
+        // with these two vectors from the camera position and the intersection of the flat plane.
+        // We know two angles, and the third is easy to compute (just 180 minus the sum of the
+        // rest). the first angle is 90 degrees + the current tilt, i.e. the tilt at the map
+        // center. The second angle is 90 - the tilt at the rotation center. Note, this doesn't work
+        // for sphere, it would have to be fixed for that.
+        const angleBetweenRotationTargetAndMapTarget =
+            ninetyRad * 2 -
+            (ninetyRad + angleBetweenNormals - currentTiltRotationCenter + ninetyRad + currentTilt);
+
+        // This uses the same formula as above, just we change `currentTilt` to the maxTiltAngle,
+        // i.e we want to find the greatest angle at the rotation target that gives us the max
+        // angle at the map center target.
+        const maxRotationTargetAngle =
+            ninetyRad * 2 - angleBetweenRotationTargetAndMapTarget - ninetyRad - maxTiltAngle;
+
+        // We compute the max tilt based on the angle of the inner triangle, note, for sphere
+        // projection this is just an approximation, in order to compute this exactly, we would
+        // need to shift the camera to the new position, compute the new normal at the map center
+        // and use this to then compute the angleBetweenNormals, it would then actually need to be
+        // done iteratively, because when shifting the camera, it is possible that we move it to a
+        // position which is outside the max tilt given the new normal. Because the `deltaTilt` is
+        // generally small, the difference in normals should be for most cases very small and the
+        // extra computation required to compute such a small value isn't worth it. The value we
+        // compute is anyway conservative, so we don't risk computing something that will put the
+        // camera outside the valid range.
+        const maxTilt = ninetyRad + angleBetweenNormals - maxRotationTargetAngle;
```

Comment:
I'm having trouble understanding what you're doing here.
The last time we discussed we agreed on just limiting the tilt at the rotation center to 89 and at the map target to maxTilt. Did you a find a problem with that solution?

Anyway, I'm not sure which triangle you mention here. I think you mean the triangle CRT, being C = camera, R = rotation target, T = map target. If that's the case, the you make the following angle calculations:

angle at T (CTR) = 90 + currentTilt
angle at R (CRT) = 90 - currentTiltRotationCenter

I don't understand those calculations.

Reply:
Yes, the other solution works in Planar, but I was getting issues with the sphere case, and I spent a few days trying to find an adequate solution, but never could, so I decided to write the algorithm again  from scratch.

Reply:
> and at the map target to maxTilt

It isn't actually limiting to the maxTilt, it just looks like it, we were applying the deltaTilt around the rotation target, and not the map target. As such it is easy to get the map target past the horizon, for example, if your rotation center is below the map target, then you have less tilt, say the map target has tilt 45, and the tilt at rotation center is 30, which gives us 15 more degrees of freedom until we reach the max tilt, hence it can easily hit the horizon (with both values < 89).

Reply:
I will reword using the terminology you suggested above.